File: code/bin/hitcounter.ts
import cdk = require('@aws-cdk/cdk');
import lambda = require('@aws-cdk/aws-lambda');
import dynamodb = require('@aws-cdk/aws-dynamodb');

export interface HitCounterProps {
  /** the function for which we want to count url hits **/
  downstream: lambda.FunctionRef;
}

export class HitCounter extends cdk.Construct {
  /** allows accessing the counter function */
  public readonly handler: lambda.Function;

  constructor(parent: cdk.Construct, id: string, props: HitCounterProps) {
    super(parent, id);

    const table = new dynamodb.Table(this, 'Hits');
    table.addPartitionKey({ name: 'path', type: dynamodb.AttributeType.String });

    this.handler = new lambda.Function(this, 'HitCounterHandler', {
      runtime: lambda.Runtime.NodeJS810,
      handler: 'hitcounter.handler',
      code: lambda.Code.directory('lambda'),
      environment: {
        DOWNSTREAM_FUNCTION_NAME: props.downstream.functionName,
        HITS_TABLE_NAME: table.tableName
      }
    });

    // grant the lambda role read/write permissions to our table
    table.grantReadWriteData(this.handler.role);

    // grant the lambda role invoke permissions to the downstream function
    props.downstream.grantInvoke(this.handler.role);
  }
}

File: code/java/src/main/java/com/myorg/HitCounter.java
package com.myorg;

import java.util.HashMap;
import java.util.Map;

import software.amazon.awscdk.Construct;
import software.amazon.awscdk.services.dynamodb.Attribute;
import software.amazon.awscdk.services.dynamodb.AttributeType;
import software.amazon.awscdk.services.dynamodb.Table;
import software.amazon.awscdk.services.lambda.Code;
import software.amazon.awscdk.services.lambda.Function;
import software.amazon.awscdk.services.lambda.FunctionProps;
import software.amazon.awscdk.services.lambda.Runtime;

public class HitCounter extends Construct {
  private final Function handler;
  private final Table table;

  public HitCounter(Construct scope, String id, HitCounterProps props) {
    super(scope, id);

    this.table = new Table(this, "Hits");
    this.table.addPartitionKey(Attribute.builder().withName("path").withType(AttributeType.String).build());

    Map<String, Object> environment = new HashMap<>();
    environment.put("DOWNSTREAM_FUNCTION_NAME", props.getDownstream().getFunctionName());
    environment.put("HITS_TABLE_NAME", this.table.getTableName());
    this.handler = new Function(this, "HitCounterHandler", FunctionProps.builder().withRuntime(Runtime.NODE_J_S810)
        .withHandler("hitcounter.handler").withCode(Code.asset("lambda")).withEnvironment(environment).build());

    // grant the lambda role read/write permissions to our table
    this.table.grantReadWriteData(this.handler.getRole());
    // grant the lambda role invoke permissions to the downstream function
    props.getDownstream().grantInvoke(this.handler.getRole());
  }

  /**
   * @return the counter function
   */
  public Function getHandler() {
    return handler;
  }

  /**
   * @return the hit counter table
   */
  public Table getTable() {
    return this.table;
  }
}
